assets/js/hooks/session/cursor_history.js

```diff
@@ -0,0 +1,123 @@
+/**
+ * Allows for recording a sequence of focused cells with the focused line
+ * and navigate inside this stack.
+ */
+export default class CursorHistory {
+  /** @private */
+  entries = [];
+
+  /** @private */
+  index = -1;
+
+  /**
+   * Adds a new cell to the stack.
+   *
+   * If the stack length is greater than the stack limit,
+   * it will remove the oldest entries.
+   */
+  push(cellId, line, offset) {
+    const entry = { cellId, line, offset };
+
+    if (this.isSameCell(cellId)) {
+      this.entries[this.index] = entry;
+    } else {
+      if (this.entries[this.index + 1] !== undefined) {
+        this.entries = this.entries.slice(0, this.index + 1);
+      }
+
+      this.entries.push(entry);
+      this.index++;
+    }
+
+    if (this.entries.length > 20) {
+      this.entries.shift();
+      this.index--;
+    }
+  }
+
+  /**
+   * Removes all matching cells with given id from the stack.
+   */
+  removeAllFromCell(cellId) {
+    // We need to make sure the last entry from history
+    // doesn't belong to the given cell id that we need
+    // to remove from the entries list.
+    let currentEntryIndex = this.index;
+    let currentEntry = this.entries[currentEntryIndex];
+
+    while (currentEntry.cellId === cellId) {
+      currentEntryIndex--;
+      currentEntry = this.entries[currentEntryIndex];
+    }
+
+    this.entries = this.entries.filter((entry) => entry.cellId !== cellId);
+    this.index = this.entries.lastIndexOf(currentEntry);
```

Comment:
`currentEntry` could be `undefined` (for example, if all entries have this cell id). It's also possible the same entry is at multiple positions in the stack.

What we can do is, iterate cells from index 0 to `<= this.index` and count how many cells have `cellId`, then we subtract that from the current index (if the index turns out to be `-1`, an the stack is not empty, we set it to 0?).